### PR TITLE
e2e: ginkgo timeouts: use context provided by ginkgo

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Ginkgo can tell code to stop running by canceling the context that Ginkgo provides to callbacks when those accept one. This can be used to stop immediately when aborting manually via CTRL-C and to clean up properly in case of a timeout, because cleanup code then runs after the main test has stopped with a new context.

#### Special notes for your reviewer:

I started eliminating context.TODO in provisioning.go and then branched out from there: any function which had context.TODO needed an explicit context. The minimal goal for this PR is to clean up all code in test/e2e/framework because then further PRs probably can be more localized.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

**Comment by aojea** [comment]
is this the one you are testing the new contexts?

**Comment by pohly** [comment]
Yes.

**Comment by pohly** [comment]
Test failures are intentional - I broke one test to see how timeouts get reported.

**Comment by pohly** [comment]
I am now at a point where `context.TODO` has been removed from all files that I touch in this PR, i.e. I've reached some kind of closure.

More work can be found with the grep commands in the last commit message, though, so further PRs would be needed if we merge this PR as-is.

**Comment by spiffxp** [comment]
/cc
Looks this builds on top of https://github.com/kubernetes/kubernetes/pull/113367

**Comment by spiffxp** [comment]
This is gigantic.  I'm making my through it assuming it is nearly ready to go, but the number of job failures is a little concerning.  At a minimum pull-kubernetes-node-e2e-containerd failing needs to be addressed (haven't had time to look into why it's failing but it failed a retest)

Is there any part of this you could break down so it doesn't get caught in rebases?

**Review by spiffxp — COMMENTED** [review]
I'm not going to be around to re-review this tomorrow. In principal I'm OK with these changes, I just think it's a lot to do all at once. Given that code freeze / test freeze are coming up I can't help but wonder if there's a way to shard these up or break into more commits for easier review.

**Comment by pohly** [comment]
> This is gigantic. I'm making my through it assuming it is nearly ready to go, but the number of job failures is a little concerning.

That's partly because I picked several pre-merge jobs and manually ran them although they are normally not started. Several of those jobs seem to be just plain broken, independently of this PR.

> At a minimum pull-kubernetes-node-e2e-containerd failing needs to be addressed (haven't had time to look into why it's failing but it failed a retest)

I'll check.

> Is there any part of this you could break down so it doesn't get caught in rebases?

Not really. Once I started requiring a context parameter in e2e/framework, it bubbled up into pretty much everything. I might have been able to stop at some point half the way, but I can't get back to that state anymore.

**Comment by pohly** [comment]
One possibility would be to wait with merging this until after code freeze and before test freeze. Then we don't make everyone else's life miserable because their new tests no longer work while they try to hit code freeze. We may still have conflicts when new tests get added after code freeze, but probably less than now.

Rebasing isn't too bad - I just take whatever chunk comes from master and fix the compile errors.

File: test/e2e/framework/test_context.go [hunk 1]
```diff
@@ -539,7 +539,9 @@ func writeJUnitReport(report ginkgo.Report) {
 		// Remove details for any spec that hasn't failed. In Prow,
 		// the test output captured in build-log.txt has all of this
 		// information, so we don't need it in the XML.
-		if specReport.State != types.SpecStateFailed {
+		switch specReport.State {
+		case types.SpecStateSkipped, types.SpecStatePassed:
```
[the remark this answers is not in the document]
Comment (pohly) [reply test/e2e/framework/test_context.go:546]: Sorry, old habits... fixed.

File: test/e2e/framework/ginkgowrapper.go [hunk 2]
```diff
@@ -18,15 +18,47 @@ package framework
...
 )
 
+var errInterface = reflect.TypeOf((*error)(nil)).Elem()
+
+// IgnoreNotFound can be used to wrap an arbitrary function in a call to
+// [ginkgo.DeferCleanup]. When the wrapped function returns an error that
+// `apierrors.IsNotFound` considers as "not found", the error is ignored
+// instead of failing the test during cleanup. This is useful for cleanup code
+// that just needs to ensure that some object does not exist anymore.
+func IgnoreNotFound(in any) any {
```
[the remark this answers is not in the document]
Comment (pohly) [reply test/e2e/framework/ginkgowrapper.go:35]: I can add it in a separate commit with a better explanation why it is needed: several `ginkgo.DeferCleanup` calls invoke a delete operation and are not supposed to fail when the object that needs to be deleted is already gone.

[the remark this answers is not in the document]
Comment (pohly) [reply test/e2e/framework/ginkgowrapper.go:35]: Actually, I already added it in `e2e framework: support ignoring "not found" errors during DeferCleanup`, but that commit is not as clean as it should be. I'll fix that.

File: test/e2e/framework/framework.go [hunk 3]
```diff
@@ -268,6 +268,9 @@ func (f *Framework) dumpNamespaceInfo() {
 	if !TestContext.DumpLogsOnFailure {
 		return
 	}
+	if f.DumpAllNamespaceInfo == nil {
+		return
+	}
```
[the remark this answers is not in the document]
Comment (pohly) [reply test/e2e/framework/framework.go:273]: This needs to be in its own commit. It fixes a nil pointer access in the output unit test where DumpAllNamespaceInfo isn't set.

**Comment by pohly** [comment]
The  pull-kubernetes-node-e2e-containerd was caused by a misplaced `ctx` parameter in a `ginkgo.Context` callback. This is not valid, which caused the suite to fail during startup. All of those container node calls get executed, so it's not possible to miss such a mistake.

**Comment by spiffxp** [comment]
>One possibility would be to wait with merging this until after code freeze and before test freeze.

That sounds reasonable, ping once you've rebased after code freeze

File: test/e2e/framework/framework.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (aojea) [reply test/e2e/framework/framework.go:273]: yeah, I hit this the other day here https://github.com/kubernetes/kubernetes/pull/113850

File: test/e2e/framework/ginkgowrapper.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (aojea) [reply test/e2e/framework/ginkgowrapper.go:35]: this is super complex :/ 
do we have many of these failures?

File: test/e2e/storage/drivers/csi-test/driver/driver.go [hunk 4]
```diff
@@ -157,7 +157,7 @@ func goServe(server *grpc.Server, wg *sync.WaitGroup, listener net.Listener, sta
 		started <- true
 		err := server.Serve(listener)
 		if err != nil {
-			panic(err.Error())
```
Comment (aojea) [inline test/e2e/storage/drivers/csi-test/driver/driver.go:160]: it seems that this should be using the `defer ginkgo.GinkgoRecover()`

File: hack/conformance/check_conformance_test_requirements.go [hunk 5]
```diff
@@ -30,7 +30,7 @@ import (
 )
 
 const (
-	//e.g. framework.ConformanceIt("should provide secure master service ", func() {
+	//e.g. framework.ConformanceIt("should provide secure master service ", func(ctx context.Context) {
```
Comment (aojea) [inline hack/conformance/check_conformance_test_requirements.go:33]: why don't you add a `hack/verify-ginkgo-context.sh`and `hack-update-ginkgo-context.sh` so we are consistent with this ?
It also will give us more confident in this change having the scripts in repo

File: test/e2e/apimachinery/apiserver_identity.go [hunk 6]
```diff
@@ -37,7 +37,7 @@ import (
 	admissionapi "k8s.io/pod-security-admission/api"
 )
 
-func getControlPlaneHostname(node *v1.Node) (string, error) {
+func getControlPlaneHostname(ctx context.Context, node *v1.Node) (string, error) {
```
Comment (aojea) [inline test/e2e/apimachinery/apiserver_identity.go:40]: I suggest to merge this commit in a follow up PR, so we have some time to observe the behavior of the CI.

File: test/e2e/framework/framework.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (pohly) [reply test/e2e/framework/framework.go:273]: There are several AfterEach callbacks that use variables set in BeforeEach or during tests, without checking that the initialization actually completed. Using DeferCleanup right after initialization avoids this pitfall, we just can't use it here and thus have to check.

File: test/e2e/framework/ginkgowrapper.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (pohly) [reply test/e2e/framework/ginkgowrapper.go:35]: There were a few, usually coming from tests that had a `defer func() { ... Delete ... }` without return code checking when those tests also deleted the object as part of normal testing.

File: test/e2e/storage/drivers/csi-test/driver/driver.go (hunk 4 above)
Comment (pohly) [reply test/e2e/storage/drivers/csi-test/driver/driver.go:160]: That would have made the panic show up as a Ginkgo failure instead of just being recorded in the log output. Panics shouldn't happen in this goroutine, therefore I am not sure whether it's worth it.

File: hack/conformance/check_conformance_test_requirements.go (hunk 5 above)
Comment (pohly) [reply hack/conformance/check_conformance_test_requirements.go:33]: Not all callbacks need a context, only those that call functions which need a context. In practice, almost all tests will need it because they communicate with a cluster, which implies that they could block and thus should use a context. But should we make it a requirement for all tests?

There may be some value in it. Developers might be tempted to write a test like this because they don't know better:
```
ginkgo.It("...", func() {
    ctx := context.TODO() // For framework helper functions below.
   ...
    e2epod.WaitForPod(ctx, ...)
})
```

The sed invocation isn't perfect, though. A few tests were defined with
```
ginkgo.It("...",
   func() {
   }
)
```

Because sed expected the `func` on the same line as `ginkgo.It`, they weren't updated. Should the scripts perhaps enforce that they are on the same line? Tests that use two lines need to indent the function body one level deeper than tests that use one line, and it is inconsistent.

File: test/e2e/apimachinery/apiserver_identity.go (hunk 6 above)
Comment (pohly) [reply test/e2e/apimachinery/apiserver_identity.go:40]: "This commit" = "[e2e: use context from Ginkgo](https://github.com/kubernetes/kubernetes/pull/112923/commits/8838fc4314822ae079132486bb0c056c8ff59e3a)"?

But then the rest of the changes are pretty harmless (I know, famous last words...). How long would you want to test those before doing the real change?

File: hack/conformance/check_conformance_test_requirements.go (hunk 5 above)
Comment (pohly) [reply hack/conformance/check_conformance_test_requirements.go:33]: I just checked, there are some places where sed is adding context parameters unnecessarily:
```
diff --git a/test/e2e/dra/dra.go b/test/e2e/dra/dra.go
index fc7529e682c..24aa9301249 100644
--- a/test/e2e/dra/dra.go
+++ b/test/e2e/dra/dra.go
@@ -61,7 +61,7 @@ var _ = ginkgo.Describe("[sig-node] DRA [Feature:DynamicResourceAllocation]", fu
                nodes := NewNodes(f, 1, 1)
                driver := NewDriver(f, nodes, networkResources) // All tests get their own driver instance.
                b := newBuilder(f, driver)
-               ginkgo.It("registers plugin", func() {
+               ginkgo.It("registers plugin", func(ctx context.Context) {
                        ginkgo.By("the driver is running")
                })
```

Here all the relevant work was done in `BeforeEach`, the test just exists to have something simple which exercises that and nothing else.

The other cases are `test/e2e/framework/log_test.go` and `test/e2e/framework/skipper/skipper_test.go` which are simpler when the callbacks don't use a context.

All of these could be converted, I just wanted to share my observation.

Comment (aojea) [reply hack/conformance/check_conformance_test_requirements.go:33]: Tim wanted to have this https://github.com/quasilyte/go-ruleguard that will allow us to write rules IIUIC :( ... but yeah, it has to be something more fancy that sed, it should understand the function signature

File: test/e2e/apimachinery/apiserver_identity.go (hunk 6 above)
Comment (aojea) [reply test/e2e/apimachinery/apiserver_identity.go:40]: some soak time, that is better than doing a full revert later or start bisecting ... it seems that if we do this today we have to merge the other tomorrow https://www.kubernetes.dev/resources/release/

File: hack/conformance/check_conformance_test_requirements.go (hunk 5 above)
Comment (pohly) [reply hack/conformance/check_conformance_test_requirements.go:33]: I remember that one. I think we may be able to use it through golangci-lint because there is one plugin which accepts ruleguard configuration, but I haven't checked.

I would prefer to postpone enforcement of new rules until the dust has settled. Let's do that for 1.27.

File: test/e2e/apimachinery/apiserver_identity.go (hunk 6 above)
Comment (pohly) [reply test/e2e/apimachinery/apiserver_identity.go:40]: I've split out the "easy" parts into [a separate PR](https://github.com/kubernetes/kubernetes/pull/113884) and rebased this one here on top of it so that the complex commit is at the end.

We can first merge that and then this one (they won't conflict), or go directly for this one.

File: hack/conformance/check_conformance_test_requirements.go (hunk 5 above)
Comment (aojea) [reply hack/conformance/check_conformance_test_requirements.go:33]: sure, it was a side comment, didn't mean to start fiddling with that now

File: test/e2e/apimachinery/apiserver_identity.go (hunk 6 above)
Comment (aojea) [reply test/e2e/apimachinery/apiserver_identity.go:40]: defer to @spiffxp

File: test/e2e/apimachinery/aggregator.go [hunk 7]
```diff
@@ -57,38 +57,24 @@ const (
...
-
-		if aggrclient == nil {
-			config, err := framework.LoadConfig()
-			if err != nil {
-				framework.Failf("could not load config: %v", err)
-			}
-			aggrclient, err = aggregatorclient.NewForConfig(config)
-			if err != nil {
-				framework.Failf("could not create aggregator client: %v", err)
-			}
```
Comment (aojea) [inline test/e2e/apimachinery/aggregator.go:90]: I thought already commented, this change is unrelated to the contexts, right? can we split it out?

Comment (pohly) [reply test/e2e/apimachinery/aggregator.go:90]: It's related because `ginkgo.DeferCleanup` can handle functions which take a context, i.e. this change became necessary when cleanTest was changed to require a contect.

It can be split out, but then it has to get merged before this PR.

Comment (pohly) [reply test/e2e/apimachinery/aggregator.go:90]: => https://github.com/kubernetes/kubernetes/pull/114392

File: test/e2e/apimachinery/flowcontrol.go [hunk 8]
```diff
@@ -313,7 +310,7 @@ func getPriorityLevelNominalConcurrency(c clientset.Interface, priorityLevelName
 
 // createFlowSchema creates a flow schema referring to a particular priority
 // level and matching the username provided.
-func createFlowSchema(f *framework.Framework, flowSchemaName string, matchingPrecedence int32, priorityLevelName string, matchingUsernames []string) (*flowcontrol.FlowSchema, func()) {
+func createFlowSchema(ctx context.Context, f *framework.Framework, flowSchemaName string, matchingPrecedence int32, priorityLevelName string, matchingUsernames []string) *flowcontrol.FlowSchema {
```
Comment (aojea) [inline test/e2e/apimachinery/flowcontrol.go:313]: `createPriorityLevel` has the same signature, however, you choose to use `ginkgo.DeferCleanup(` here instead of returning a `cleanup` function , I think that is better to be consistent, maybe avoid the ginkgo primitives by now?

File: test/e2e/apimachinery/webhook.go [hunk 9]
```diff
@@ -1727,21 +1699,19 @@ func registerWebhookForCustomResource(f *framework.Framework, configName string,
 	})
 	framework.ExpectNoError(err, "registering custom resource webhook config %s with namespace %s", configName, namespace)
 
-	err = waitWebhookConfigurationReady(f)
+	err = waitWebhookConfigurationReady(ctx, f)
 	framework.ExpectNoError(err, "waiting for webhook configuration to be ready")
-	return func() {
-		client.AdmissionregistrationV1().ValidatingWebhookConfigurations().Delete(context.TODO(), configName, metav1.DeleteOptions{})
-	}
+	ginkgo.DeferCleanup(framework.IgnoreNotFound(client.AdmissionregistrationV1().ValidatingWebhookConfigurations().Delete), configName, metav1.DeleteOptions{})
```
Comment (aojea) [inline test/e2e/apimachinery/webhook.go:1704]: these are the behavioural changes I'd prefer to not do at the same time, this wasn't asserting any error before ... maybe we can avoid all these `ginkgo.DeferCleanup` changes and do only the contexts in this commit?

File: test/e2e/apimachinery/webhook.go [hunk 10]
```diff
@@ -2166,19 +2132,25 @@ func registerSlowWebhook(f *framework.Framework, configName string, certCtx *cer
...
 	framework.ExpectNoError(err, "waiting for webhook configuration to be ready")
 
-	return func() {
-		client.AdmissionregistrationV1().ValidatingWebhookConfigurations().Delete(context.TODO(), configName, metav1.DeleteOptions{})
+	cleanup := func(ctx context.Context) {
+		err := client.AdmissionregistrationV1().ValidatingWebhookConfigurations().Delete(ctx, configName, metav1.DeleteOptions{})
+		if !apierrors.IsNotFound(err) {
+			framework.ExpectNoError(err)
+		}
 	}
```
Comment (aojea) [inline test/e2e/apimachinery/webhook.go:2143]: it is more verbose but it is easier to understand that the internals of `framework.IgnoreNotFound(` using reflection and type assertion

File: test/e2e/apimachinery/flowcontrol.go (hunk 8 above)
Comment (pohly) [reply test/e2e/apimachinery/flowcontrol.go:313]: Using `ginkgo.DeferCleanup` was often the right solution for dealing with contexts during test cleanups. Not using it would have been more work and the outcome would have been more complicated. I had two choices:
- change the cleanup function to accept a context, which also means changing the caller
- eliminate the cleanup function

File: test/e2e/apimachinery/webhook.go (hunk 9 above)
Comment (pohly) [reply test/e2e/apimachinery/webhook.go:1704]: Deferring the usage of `ginkgo.DeferCleanup` means that all of these lines must be touched twice.

Comment (pohly) [reply test/e2e/apimachinery/webhook.go:1704]: Let me try one thing: I'll do a pass through the PR and extract those parts that introduce `ginkgo.DeferCleanup`. It should be possible to merge those parts first.

It will take time, though, because in several places I'll then conflict with my own follow-up changes.

Comment (aojea) [reply test/e2e/apimachinery/webhook.go:1704]: that will be great, thanks, just the context only are straightforward to review and approve ...

File: test/e2e/apimachinery/flowcontrol.go (hunk 8 above)
Comment (aojea) [reply test/e2e/apimachinery/flowcontrol.go:313]: but in some places the context is provided in the constructor, so the closure should inherit it , no?

Comment (pohly) [reply test/e2e/apimachinery/flowcontrol.go:313]: Each context from Ginko is only valid in the function that it is passed into. Using the context of the `BeforeEach` or `It` also for the cleanup operation is a bug. That should never be done because the API calls in the cleanup operation then abort immediately with "context cancelled" - one reason why it is important to check the error of those calls.

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

**Comment by pohly** [comment]
/retest

Unrelated flake: https://storage.googleapis.com/k8s-triage/index.html?test=Pods%20Extended%20Pod%20Container%20lifecycle%20should%20not%20create%20extra%20sandbox%20if%20all%20containers%20are%20done

**Comment by pohly** [comment]
pull-kubernetes-e2e-gci-gce-ingress, "LoadBalancers should be able to preserve UDP traffic when server pod cycles for a LoadBalancer service on the same nodes" is an unrelated test flake:

https://storage.googleapis.com/k8s-triage/index.html?text=Failed%20to%20hit%20the%202%20UDP&test=LoadBalancers%20should%20be%20able%20to%20preserve%20UDP%20traffic%20when%20server%20pod%20cycles%20for%20a%20LoadBalancer%20service%20on%20the%20same%20nodes

**Comment by pohly** [comment]
/retest

"pull-kubernetes-e2e-capz-windows-containerd — Pod pending timeout"

**Comment by aojea** [comment]
I tried my best to get to all files but there are 418 files changes, that is impossible, most of them are mechanical changes s/context.TODO/context gingo/ and we can always revert.


I think that merging on weekend will give us some soak time to see if something breaks, so we can revert in that case

/lgtm
/approve

e2e framework should be able to stop cleanly

**Comment by onsi** [comment]
🙌
